tap_appstore/client.py

```diff
@@ -18,13 +23,138 @@
+
+    def get_report_date(self, date):
+        """Return the report date formatted according to the specific needs of the stream."""
+        return date.strftime('%Y-%m-%d')
+
+    def download_data(self, start_date, api):
+        """Set up the endpoint for the API call. Override in subclass as needed."""
+        return None
+
+    def process_record(self, record):
+        """Convert ints, floats, and date fields"""
+        if self.float_fields:
+            self.convert_fields(record, self.float_fields, float)
+
+        if self.int_fields:
+            self.convert_fields(record, self.int_fields, int)
+
+        if self.date_fields:
+            for field, format in self.date_fields.items():
+                if field in record and record[field]:
+                    record[field] = self.convert_date(record[field], format)
+
+        return record
+
+    def increment_date(self, date):
+        """Increment date by one day. Override in subclass if different increment is needed."""
+        return date + timedelta(days=1)
+
+    def update_stream_state(self, date):
+        """Update the stream state with the new date."""
+        self.stream_state['start_date'] = date.strftime('%Y-%m-%d')
+        self.logger.info(f"Updating state, new start date is {self.stream_state['start_date']}")
+
+    def get_records(self, context: dict = None):
+        """Return a generator of record-type dictionary objects."""
+        start_date = self.get_start_date(default_date='2024-01-01')
+        if not start_date:
+            logger.error("Start date could not be determined.")
+            return
+
+        start_date = start_date.replace(tzinfo=None)
+        date_limit = datetime.now().replace(tzinfo=None) - timedelta(days=2)
+        line_id = 0
+
+        while start_date <= date_limit:
+
+            try:
+                logger.info(f"report_date: {self.get_report_date(start_date)}")
+                all_data = self.download_data(start_date, self.api)
+
+                data_io = StringIO(all_data)
+
+                # Rename column names from 'App Name' -> 'app_name'
+                first_line = data_io.readline().strip()
+                fieldnames = [col.strip().replace(' ', '_').lower() for col in first_line.split('\t')]
+
+                reader = csv.DictReader(data_io, delimiter='\t', fieldnames=fieldnames)
+
+                for record in reader:
+                    line_id += 1
+                    record['_line_id'] = line_id
+                    record['_time_extracted'] = datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ')
+                    record['_api_report_date'] = self.get_report_date(start_date)
+
+                    processed_record = self.process_record(record)
+                    if processed_record is not None:
+                        yield processed_record
+
+                    yield record
+
+            except APIError as e:
+                logger.error(f'Error during download report {self.name}.\n{e}')
```

Comment:
raise an exception or we will lose data. Unless there is no report for that date.